tracex/extraction/logic/modules/module_metrics_analyzer.py

```diff
@@ -0,0 +1,110 @@
+    - correctness of timestamps
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.name = "Metrics Analyzer"
+        self.description = (
+            "Measures the output of the pipeline based on specified metrics."
+        )
+
+    @log_execution_time(Path("extraction/logs/execution_time.log"))
+    def execute(self, df, patient_journey=None):
+        super().execute(df, patient_journey)
+
+        return self.__measure_metrics(df)
+
+    def __measure_metrics(self, df):
+        """Executing the measurement of metrics. The metrics output will write on disk as a csv file.
+        The dataframe without the metrics is returned for visualization."""
+
+        metrics_df = df.copy()
+        metrics_df["activity_relevance"] = metrics_df["activity"].apply(
+            self.__rate_activity_relevance
+        )
+
+        metrics_df[
+            ["timestamp_correctness", "correctness_confidence"]
+        ] = metrics_df.apply(
+            lambda row: pd.Series(
+                self.__rate_timestamps_correctness(
+                    row["activity"], row["start"], row["end"]
+                )
+            ),
+            axis=1,
+        )
+
+        metrics_df.to_csv(
+            Path(c.output_path.joinpath("metrics.csv")), index=False, sep=","
+        )
+
+        return df
+
+    def __rate_activity_relevance(self, activity):
+        category_mapping = {
+            "No Relevance": 0,
+            "Low Relevance": 1,
+            "Moderate Relevance": 2,
+            "High Relevance": 3,
+        }
+
+        messages = [
+            {"role": "system", "content": p.METRIC_ACTIVITY_CONTEXT},
+            {
+                "role": "user",
+                "content": (
+                    f"{p.METRIC_ACTIVITY_PROMPT}\nThe bulletpoint: {activity}\n"
+                    f"The patient journey: {self.patient_journey}"
+                ),
+            },
+        ]
+
+        answer = u.query_gpt(messages)
+        for key in category_mapping:
+            if key in answer:
+                category = key
+                break
+
+        return category
+
+    def __rate_timestamps_correctness(self, activity, start, end):
+        messages = [
+            {"role": "system", "content": p.METRIC_TIMESTAMPS_CONTEXT},
+            {
+                "role": "user",
+                "content": (
+                    f"{p.METRIC_TIMESTAMPS_PROMPT}\nThe bulletpoint: {activity}\n"
+                    f"The start date related to the bulletpoint: {start}\n"
+                    f"The end date to the bulletpoint: {end}\n"
+                    f"The patient journey you should check the timestamps for the bulletpoint: {self.patient_journey}"
+                ),
+            },
+        ]
+
+        timestamp_correctness, top_logprops = u.query_gpt(
+            messages, logprobs=True, top_logprobs=1
+        )
+        linear_prop = self.__calculate_linear_probability(top_logprops[0].logprob)
```

Comment:
Why do you need an extra method here?

Reply:
For the use of logprob in future metrics an extra method that converts the log probability to a linear probability would be useful instead to convert it manually in every new metric that needs logprob.